### ISSUE TITLE
acme: (Initial) Generation of multiple certs can be triggered concurrently, causing high system load

### ISSUE THREAD

**Issue by osnyx** [NixOS/nixpkgs#232505]
### Describe the bug
When setting up TLS certificates for multiple domains via the `security.acme` module, their generation might run concurrently at the same time. With a high number of certificates, this can cause a high load on the system, potentially even rendering it unresponsive.
While the renewal service run activation timers being spread out over time prevents this load concurrency, the issue appears when the services are simultaneously activated by a system switch. This is the case when generating them initially, and also when parameters of the certs or their services are modified enough to make the NixOS switch logic consider the services to require a re-activation.
For this to happen, all the certificates need to be linked to the same acme registry account (commonly: letsencrypt account), see the analysis further below.

### Steps To Reproduce
1. Set up a large number of acme certificates, see example config below.
2. open a `journalctl -f` and (h)top to observe system behavious
3. `nixos-rebuild switch`
4. observation 1: First, only one of the certificates is generated.
5. observation 2: Once that generation process has finished, all other remaining certificates are generated concurrently. This can increase system load.
6. To reset the certificates and generate all of them again, you can do:
  - `systemctl clean --all --what=state "acme-domain*.test.fcio.net.service"`
  - `systemctl stop --all acme-finished*.target`
  - `systemctl start --all acme-finished*.target`
example config
```nix
{
# /var/lib/acme/.challenges must be writable by the ACME user
# and readable by the Nginx user. The easiest way to achieve
# this is to add the Nginx user to the ACME group.
users.users.nginx.extraGroups = [ "acme" ];

flyingcircus.services.nginx = {
  enable = true;
  virtualHosts = {
    "domain1.test.fcio.net" = {
      # Catchall vhost, will redirect users to HTTPS for all vhosts
      serverAliases = [ "*.test.fcio.net" ];
      locations."/.well-known/acme-challenge" = {
        root = "/var/lib/acme/.challenges";
      };
      locations."/" = {
        return = "301 https://$host$request_uri";
      };
    };
  };
};

security.acme.certs =
  let certdata = {
          webroot = "/var/lib/acme/.challenges";
          # Ensure that the web server you use can read the generated certs
          # Take a look at the group option for the web server you choose.
          group = "nginx";
          # Since we have a wildcard vhost to handle port 80,
          # we can generate certs for anything!
          # Just make sure your DNS resolves them.
          # Use staging ACME verification server
          server = "https://acme-staging-v02.api.letsencrypt.org/directory";
  };
  in {
  "domain1.test.fcio.net" = certdata;
  "domain2.test.fcio.net" = certdata;
  "domain3.test.fcio.net" = certdata;
  "domain4.test.fcio.net" = certdata;
  "domain5.test.fcio.net" = certdata;
  "domain6.test.fcio.net" = certdata;
  "domain7.test.fcio.net" = certdata;
  "domain8.test.fcio.net" = certdata;
  "domain9.test.fcio.net" = certdata;
  "domain10.test.fcio.net" = certdata;
  "domain11.test.fcio.net" = certdata;
  "domain12.test.fcio.net" = certdata;
};
security.acme.acceptTerms = true;
security.acme.defaults.email = "admin@example.com";
}
```

### Expected behavior
The concurrency of certificate generation shall be limited to a certain number of parallel runs, possibly configurable.

### Analysis

The reason for the behaviour described above has been introduced in PR #106857:
All the services for generating a particular domain cert depend on a target for the account:

```command
# systemctl -l cat acme-account-<hash>.target | cat
# /etc/systemd/system/acme-account-<hash>.target
[Unit]
After=acme-domain1.test.fcio.net.service
Before=acme-domain10.test.fcio.net.service acme-domain11.test.fcio.net.service acme-domain12.test.fcio.net.service acme-domain2.test.fcio.net.service acme-domain3.test.fcio.net.service acme-domain4.test.fcio.net.service acme-domain5.test.fcio.net.service acme-domain6.test.fcio.net.service acme-domain7.test.fcio.net.service acme-domain8.test.fcio.net.service acme-domain9.test.fcio.net.service
Requires=acme-domain1.test.fcio.net.service
```

So the first domain script is run alone, but afterwards all the other services are runnable and thus executed at the same time by systemd.
The individual services are activated by a `wantedBy` relationship to `multi-user.target`, as well as through their domain-specific `acme-finished-<domain>.target` that has a `wantedBy´ towards `default.target`.

### Notify maintainers

@NixOS/acme
@aanderse @flokli @andrew-d @arianvp @m1cr0man @emilazy 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

disclaimer: This issue has been investigated and reproduced on our flyingcircusio/fc-nixos platform which differs in a few details, but I am rather certain I've identified the general problem being in nixpkgs.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.109, NixOS, 22.11 (Raccoon), 22.11.2560.5eed30dd`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.1`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`
```

**Comment by osnyx** [comment]
## potential idea for fixing

In my eyes, it is necessary to limit the number of concurrent service runs. One option is mutexes, an even better implementation would be using semaphore semantics to allow a restricted number of runs in parallel.

As I have not found any solutions integrated in systemd to enforce such limit, I am currently investigating the usage of `flock` for file-based mutexes of script runs via their shared state directory. Once this is ready, I'll propose a PR for further discussion.
I am open for any further ideas, either for locking or even different approaches, so feel free to propose them.

**Comment by emilazy** [comment]
I dread seeing the ACME module grow even more bespoke complexity of this kind if at all avoidable. Would it be possible to avoid the negative effects here in a different way by using the [systemd resource control settings](https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html) to limit the CPU weight and/or quota assigned to the certificate generation so that they will yield CPU to other processes on the system without being manually serialized?

**Comment by m1cr0man** [comment]
I feel this is more of a feature right now than a bug. Other than system load, there's no reason it can't run all these renewals at once after the first (infact allowing concurrent runs is the primary reason for that account target).

In the same vein as emilazy, I wouldn't be keen to add some sort of flock logic to the renewal process given how complex it already is. We recently [pushed back](https://github.com/NixOS/nixpkgs/pull/199033) on some other changes which would have increased complexity for minimal gain for the module.

I feel systemd resource control (CPUQuota) might be sufficient here too, and I would be willing to stretch to creating a systemd slice (or whatever would work) to encapsulate all the renewals so that they collectively share the same quota to prevent ballooning when you have _lots_ of certificates. This would need further investigation, I've personally never set up quotas like this before, but I have heard it is possible.

**Comment by emilazy** [comment]
I think if the CPU weight works then it's preferable to an overall quota, as it means an unloaded system can issue all the certificates faster rather than imposing a hard limit. (But I don't know if it will work; my experience with `nice`ing processes has been mixed.)

**Comment by osnyx** [comment]
Thank you for your thoughts on this so far. While I still have some concerns about resource limits not being a sufficient counter measure – I'll lay them out below – I will build a prototype with resource limits before pursuing the lock strategy any further. Managing limited resources on a sommon slice level sounds like an approach worth investigating.

@m1cr0man 
> I feel this is more of a feature right now than a bug. Other than system load, there's no reason it can't run all these renewals at once after the first (infact allowing concurrent runs is the primary reason for that account target).

On the one hand, when we look at how new certificates are generated after a system switch, it is a feature indeed: if everything goes well and there are enough resources, the new certificates are available as quick as possible after activating the new config and thus the generator service units are not artificially throttled.
But it's not that we (read: the NixOS ACME module) do not care about concurrency issues at all: The renewal timer runs are spread out uniformly at random over time to actually avoid or limit concurrent renewal ones. This would'nt be implemented like this if we ignored the issue of concurrency.
On the other hand, right now we have a more or less linear growth of required resources for the number of certificates we generate at switch time. And in contrast to other concurrency requirements, like wanting to run multiple database services on the same machine instead of one, this is not strictly necessary, as the generation jobs could be serialised without such a high peak load. I'd even go so far to say that having a module design that does *not* become brittle when a high peak load requirement cannot be fulfilled is an equally important feature as well.

Also note there is a [rate limit on Letsencrypt's side](https://letsencrypt.org/docs/rate-limits/):
> the “new-nonce”, “new-account”, “new-order”, and “revoke-cert” endpoints on the API have an Overall Requests limit of 20 per second.

So generating significantly more than 20 certs at the same time might make parts of them fail.


@emilazy 
> by using the [systemd resource control settings](https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html) to limit the CPU weight and/or quota assigned to the certificate generation so that they will yield CPU to other processes on the system without being manually serialized

You two have been mainly talking about CPU constraints. But as you can see below, CPU is not the only contended resource affected by concurrent lego runs. And while with weights there is indeed a *yield* mechanism for CPU contention, there isn't one for memory. There I'd need to impose hard limits.

![grafana resource dashboard](https://github.com/NixOS/nixpkgs/assets/104593071/61802508-c1ff-47bd-9540-bd4911147e98)

My **main concern with the resource limit approach** is how the individual lego **processes can deal with being CPU-throttled or denied more requested memory**. Will some of them just crash? Killed by the OOM? Run into timeouts on the ACME registry side? In all of these cases, at least some of the certificates won't be generated successfully. I'd prefer a throttled, but successful generation of certs over that any time.

Also, which resources do we need to limit and to *which defaults*?

- CPU:
  - `StartupCPUWeight` for the slice is the most convenient due to its dynamic yield behaviour, but is it sufficient?
  - `CPUQuota` can be given in `%`, but is still depending on the number of cores. 100% is a full core. Which default value is tolerable both for all machines, no matter single- or multi-core?
- memory:
  - exceeding `MemoryHigh` throttles the particular process(es), but they can still allocate more memory, just much slower. Does this cause timeouts?
  - exceeding `MemoryMax` invokes the OOM killer, causing cert generation to fail
  - again, which percentage of memory is an acceptable default for all kinds of machines?
- IO limits/ network limits are hopefully not required

If you have any good experience in resource-limiting services, feel free to point out where my concerns are misplaced.

**Comment by osnyx** [comment]
One further comment on my lock-based idea: I had only intended to start with the mutually-excluding `flock` as a first try. Having a semaphore lock solution, allowing for a limited number of concurrent runs, is a far better approach and potential improvement. I just need to find a good semaphore solution that does not leave stale locks around after process crash. `flock` appears to unlock automatically once a process crashes and thus the file descriptors used for locking are released.

**Comment by osnyx** [comment]
**TL;DR:** I ran a heavily-restricted (120MiB memory limit) concurrent generation of 25 certs. 2 of them failed. Is this a relevant edge case of too few resources we want to neglect and allow failures?

As an experiment, I ran an experiment with 25 concurrent certs, all running within the same slice restricted by the following resource limits:

```nix
      systemd.slices.nixosAcme.sliceConfig = {
        StartupCPUWeight = 20;  # default is 100, range is 1-1000, lower means fewer resources.
        # do wen need CPUQuota as well?
        MemoryHigh = "100M";
        MemoryMax = "120M";
      };
```

The good thing is, that the system did not appear to become overloaded. Note that I did not (yet) run another actual load-generating service at the same time that competes for the same resources, but despite the rather strict memory limit of 100MiB – until processes are slowed down – or 120MiB – until they are OOM-killed – the acme services were able to fully utilise the single (virtual) CPU core.

![Grafana VM stats](https://github.com/NixOS/nixpkgs/assets/104593071/01acba7c-8e7d-4997-b973-fe466bc3709e)


The not-so-good result is, that **two of the certificates failed due to timeouts**:

```
root@test71 ~ # systemctl status acme-domain16.test.fcio.net.service
× acme-domain16.test.fcio.net.service - Renew ACME certificate for domain16.test.fcio.net
     Loaded: loaded (/etc/systemd/system/acme-domain16.test.fcio.net.service; linked; preset: enabled)
     Active: failed (Result: exit-code) since Mon 2023-06-12 19:39:58 CEST; 2min 0s ago
TriggeredBy: ● acme-domain16.test.fcio.net.timer
    Process: 3658033 ExecStart=/nix/store/a43d0falm2w5ic2ca3nwb4zif6y9hjn3-unit-script-acme-domain16.test.fcio.net-start/bin/acme-domain16.test.fcio.net-start (code=exited, status=10)
   Main PID: 3658033 (code=exited, status=10)
         IP: 4.1K in, 1.0K out
        CPU: 402ms
```

```text
Jun 12 19:38:48 test71 acme-domain16.test.fcio.net-start[3658033]: + lego --accept-tos --path . -d domain16.test.fcio.net --email admin@flyingcircus.io --key-type ec256 --http --http.webroot /var/lib/acme/.challenges --server https://acme-staging-v02.api.letsencrypt.org/directory run
Jun 12 19:39:28 test71 acme-domain16.test.fcio.net-start[3658085]: 2023/06/12 19:39:28 Could not create client: get directory at 'https://acme-staging-v02.api.letsencrypt.org/directory': Get "https://acme-staging-v02.api.letsencrypt.org/directory": net/http: TLS handshake timeout
Jun 12 19:39:28 test71 acme-domain16.test.fcio.net-start[3658033]: + echo Failed to fetch certificates. This may mean your DNS records are set up incorrectly. Selfsigned certs are in place and dependant services will still start.
Jun 12 19:39:28 test71 acme-domain16.test.fcio.net-start[3658033]: Failed to fetch certificates. This may mean your DNS records are set up incorrectly. Selfsigned certs are in place and dependant services will still start.
Jun 12 19:39:28 test71 acme-domain16.test.fcio.net-start[3658033]: + exit 10
```

Not that the DNS-records are actually fine and re-running the service succesfully generates a certificate. The `net/http: TLS handshake timeout` is a result of the applied resource restrictions instead.
The failing cert is not being retried immediately and thus remains missing until the next daily timer run.


So what does this mean for the feasability of resource restrictions for ACME service runs?
To be fair, 25 concurrent certificates combined with just 100-120MiB of RAM might be not that common. More acceptable limits like 20% (MemoryHigh) and 25% (MemoryMax) on the same machine (2GiB RAM + 2GiB swap) make all certificate services succeed without any failures.
But on the other hand this shows that resource limits are at best a partial mitigation for the issue of too many concurrent runs, while pursuing the approach of (partially) serialising service execution order with locks will probably always work.

What do you think of these result?

**Comment by arianvp** [comment]
Other idea which doesn't involve any locks

We add a `OnActiveSec=1s` to the `.timer` unit. This causes the service to start 1s (+ `RandomizedDelaySec` which we already have) after the `.timer` is started.

We then modify the activation script to not start the services, but only the `timer`. This will cause the timer to start; and after 1 second start all the services with a randomised delay.

**Comment by m1cr0man** [comment]
Thank you for the thorough research @osnyx :) The graphs are really interesting to see, and so are the failures caused by resource constraining the renewal services. Indeed it seems that resource constraining would be both insufficient and detrimental to reliability. Beyond that, the LetsEncrypt rate limits you highlighted are a real issue too, and we do tend to solve for those where possible.

> We add a `OnActiveSec=1s` to the `.timer` units. This causes the service to start 1s (+ `RandomizedDelaySec` which we already have) after the `.timer` is started.

This is promising, but wouldn't services with Wants/After dependencies on a certificate still trigger a start on activation? For example, [apache](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/services/web-servers/apache-httpd/default.nix#L777-L778) or [nginx](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/services/web-servers/nginx/default.nix#L1181-L1182).

Maybe it is sufficient for us to have a dependency on the acme-selfsigned-*.service for each cert (for each web server), and then use the timer to distribute when each one executes, but I can't work out the systemd behaviour in my head, it would need to be tested.

> The reason for the behaviour described above has been introduced in PR https://github.com/NixOS/nixpkgs/pull/106857

This gave me an idea.. Would it be reasonable for us to simply chain the certificates together with `After=`? This way, each waits for the previous one to run. Afaik systemd is smart enough to run the account-creating certificate first anyway. `After` wouldn't create cascading failures either, it would just set the ordering. The downside to this is that a configuration switch could take a ridiculous amount of time. What we would need then is a target or something which does not delay activation but does start the services.

There does come a point where the complexity of solving a problem with systemd alone is not reasonable, and I fear we're close to that here. If we do opt for a locking solution, I propose we go all-in and make it a semaphore so that some level of concurrency can remain for optimal performance.

**Comment by osnyx** [comment]
Thanks for your ideas so far. The approach of @arianvp is definitely an interesting and possible way of approaching this. One concern though is that the `RandomizedDelaySec` are distributed within an interval of 24 hours. Given that this overload phenomenon mostly appears at initial generation of certificates, having to wait several hours *after* the configuration switch might not be an UX impact we want to accept.

The `After=` chaining idea of @m1cr0man also has its perks and might even be favourable over a mutex-locking approach with paralellism of max. 1. Because when leaving the ordering and execution of jobs to systemd, each individual renewal service only runs when it is actually free to run. With lockin *inside the start script* of each service, we face the danger of running into startupt timeouts as even services not holding the lock and thus waiting already count as active from systemd's perspective.

Nonetheless my next attempts will still focus on locking – but with a paralellism > 1. The improved overall runtime and configuration switch time of parallel runs hopefully outweighs the possibility of service timeouts. For complexity reasons I'll probably won't implement real semaphors, but instead will use multiple mutex locks and a static assignment of services to these locks.
You can await my PR on this.

**Comment by emilazy** [comment]
There is functionality being discussed upstream in systemd that would be very relevant to this usecase: https://github.com/systemd/systemd/issues/28075

I'm especially hesitant to increase the amount of internal complexity given that, since it may become obsolete and need ripping out again sooner than later.

**Comment by osnyx** [comment]
@emilazy Thanks for pointing me towards this. Unfortunately the existing PR might have a different direction right now. https://github.com/systemd/systemd/issues/28075#issuecomment-1611025617

Let's wait on how this develops for a few days, depending on this I'll decide on whether to get started with a custom locking solution instead.

**Comment by osnyx** [comment]
We now have two different implementations for tackling this issue:
- #244511 using `flock` on groups of services to prevent them from running their script concurrently
- #246665 creates groups of `After=` systemd ordering restrictions to prevent them from activating concurrently

Now we need to decide on which of these approaches to finally adopt, refine, and then merge (for the `flock` case I still need to write tests, which I'd be willing to do).

TL;DR: I could probably live with both of them. While I find the wider concurrency limit enforcement of `flock` a bit preferrable, I'd agree with the choice the NixOS/acme maintainers team feels more comfortable about.

Below a broader discussion of some details and their implications (everything IMHO, feel free to disagree or correct):

## scope

With the `flock` solution, the concurrency limits are part of the main service script and thus are always enforced, no matter how the services are started.

With the `After` solution, the concurrency limits are only enforced for some ways of activating the services. The most important case of switch-to-configuration is covered reliably. What is not covered is the service activation via the renew timers. As these timers are randomly spread out over 24 hours by default, that is acceptable IMHO.
One case where the concurrency limits are not met as well is when manually starting the `acme-<domain>.service`s or `acme-finished*.target`s, as long as they've been successfully activated once. I've observed more concurrent lego processes running than the configured limit. In this case, we can of course discuss whether this is some deliberately-shooting-yourself-in-the-foot which does not need to be covered, but we at least better be aware of this.

## complexity

Personally, I see both implementations as similarly complex. The hard thing to understand is the creation of inter-service relationships from Nix code. In the case of the `flock` solution, the management of lock files blows up the code additionally, but IMHO does not include much complexity or moving parts.

This is the main part where you, the acme maintainers, need to decide what is more maintainable *to you*.

Both cases should be resilient against blocking service activation forever. With the `After=` solution, I manually checked the behaviour of systemd with failed referenced units. And for the `flock` solution, locks are automatically released when the script exits or crashes.

## visibility

As the `After=` approach utilises existing systemd machinery and puts the concurrency limitation outside of the actual service script execution, the service statuses more thoroughly reflects whether the actual service generation is running or not.
With the `flock` approach, services are `activating` even when they're still blocked on waiting for their lock. You'd need to check the journal to see what's going on.

**Comment by osnyx** [comment]
Any updates on this, how can we proceed here?

I'd still like to wait on feedback to decide which of the 2 PRs to fine-tune and finish, both still need minor adjustments.

I'll be at NixCon, so the hackday there could be an opportunity for finishing this in case some of you attend as well, but there's also ample time to wrap this up in the next few weeks.
